examples/simple_fwd_tb/forward_tb.c

```diff
@@ -0,0 +1,325 @@
+                                        RTE_LOG(INFO, APP, "Unknown option `-%c'.\n", optopt);
+                                else
+                                        RTE_LOG(INFO, APP, "Unknown option character `\\x%x'.\n", optopt);
+                                return -1;
+                        default:
+                                usage(progname);
+                                return -1;
+                }
+        }
+
+        if (!dst_flag) {
+                RTE_LOG(INFO, APP, "Simple Forward with Token Bucket NF requires destination flag -d.\n");
+                return -1;
+        }
+
+        return optind;
+}
+
+/*
+ * This function displays stats. It uses ANSI terminal codes to clear
+ * screen when called. It is called from a single non-master
+ * thread in the server process, when the process is run with more
+ * than one lcore enabled.
+ */
+static void
+do_stats_display(struct rte_mbuf *pkt) {
+        const char clr[] = {27, '[', '2', 'J', '\0'};
+        const char topLeft[] = {27, '[', '1', ';', '1', 'H', '\0'};
+        static uint64_t pkt_process = 0;
+        struct ipv4_hdr *ip;
+
+        pkt_process += print_delay;
+
+        /* Clear screen and move to top left */
+        printf("%s%s", clr, topLeft);
+
+        printf("PACKETS\n");
+        printf("-----\n");
+        printf("Port : %d\n", pkt->port);
+        printf("Size : %d\n", pkt->pkt_len);
+        printf("N°   : %" PRIu64 "\n", pkt_process);
+        printf("\n\n");
+
+        ip = onvm_pkt_ipv4_hdr(pkt);
+        if (ip != NULL) {
+                onvm_pkt_print(pkt);
+        } else {
+                printf("No IP4 header found\n");
+        }
+}
+
+static int
+packet_handler_tb(struct rte_mbuf *pkt, struct onvm_pkt_meta *meta,
+                  __attribute__((unused)) struct onvm_nf_local_ctx *nf_local_ctx) {
+        static uint32_t counter = 0;
+        uint64_t tokens_produced;
+
+        /* Generate tokens only if `tb_tokens` is insufficient */
+        if (tb_tokens < pkt->pkt_len) {
+                cur_cycles = rte_get_tsc_cycles();
+                /* Wait until sufficient tokens are available */
+                while ((((cur_cycles - last_cycle) * tb_rate * 1000000) / rte_get_timer_hz()) + tb_tokens <
+                       pkt->pkt_len) {
+                        cur_cycles = rte_get_tsc_cycles();
+                }
+                tokens_produced = (((cur_cycles - last_cycle) * tb_rate * 1000000) / rte_get_timer_hz());
+                /* Update tokens to a max of tb_depth */
+                if (tokens_produced + tb_tokens > tb_depth) {
+                        tb_tokens = tb_depth;
+                } else {
+                        tb_tokens += tokens_produced;
+                }
+
+                last_cycle = cur_cycles;
+        }
+
+        if (++counter == print_delay) {
+                do_stats_display(pkt);
+                counter = 0;
+        }
+
+        tb_tokens -= pkt->pkt_len;
```

Comment:
@rohit-mp Can pkt->pkt_len be greater than tb_tokens? If so since both are **unsigned ints** this will result in a large number.